### PR TITLE
Added precipitation visualisations, and formatted Weather page.

### PR DESCRIPTION
**Pull request by MurasatoNaoya**
Added the fourth and final visualisation for the Weather page, for precipitation likelihood (%) and precipitation level (mm). Both weather indicators are plotted on separate x-axes, on a y-axis of time. All weather visualisations are also formatted into two separate columns holding two visualisations each.

### CODE REVIEW

File: dashboard/dashboard.py [hunk 1]
```diff
@@ -369,64 +406,69 @@ def get_current_datetime() -> datetime:
...
+    #     'y': visibility_df['body_name'],     
+    #     'z': visibility_df['visibility_m']    
+    # })
+
+
+    # heatmap = alt.Chart(source).mark_rect().encode(
+    #     x=alt.X('x:O', title='Date', axis=alt.Axis(labels=False, ticks=False)),
+    #     y=alt.Y('y:O', title='Celestial Body'),
+    #     color=alt.Color('z:Q',
+    #                     scale=alt.Scale(domain=[0, 30000],
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:462]: Remove commented code

File: dashboard/load_dashboard_data.py [hunk 2]
```diff
@@ -69,13 +71,95 @@ def load_forecasts_by_county_name(conn:Psycopg2Connection) -> pd.DataFrame:
...
+    OR EXTRACT(HOUR FROM f.at) <= 5 
+    GROUP BY r.region_id, r.region_name, f.at), br as
+    (SELECT b.body_name,r.region_id, region_name,ba.at 
+    FROM body AS b 
+    JOIN  body_assignment AS ba
+    USING (body_id) JOIN region AS r 
+    USING (region_id) WHERE b.body_name NOT ILIKE 'Sun') 
+    SELECT af.region_id, br.region_name,br.body_name,br.at,af.avg_cloud, af.avg_vis 
+    FROM af JOIN br ON af.region_id = br.region_id AND br.at 
+    = af.at;
```
Comment (AlexOBarnes) [inline dashboard/load_dashboard_data.py:98]: What a beautiful query

File: dashboard/load_dashboard_data.py [hunk 3]
```diff
@@ -69,13 +71,95 @@ def load_forecasts_by_county_name(conn:Psycopg2Connection) -> pd.DataFrame:
...
+        '''
+        Calculates the starwatch coefficient based on visibility, distance, 
+        and cloud coverage using a weighted formula and sigmoid normalization.
+        '''
+        visibility_fraction = visibility_m / distance_km
+        visibility_weighted = visibility_fraction ** 0.10
+        adjusted_cloud_coverage = 1 - (cloud_coverage_percent / 100) ** 3
+        coefficient = visibility_weighted * adjusted_cloud_coverage
+        sigmoid_coefficient = 1 / (1 + np.exp(-10 * (coefficient - 0.5)))
+        return max(0, min(sigmoid_coefficient, 1))
```
Comment (AlexOBarnes) [inline dashboard/load_dashboard_data.py:149]: Patent this

**Review by AlexOBarnes — CHANGES_REQUESTED** [review]
Need to remove commented out code

File: dashboard/dashboard.py (hunk 1 above)
Comment (olliejth) [reply dashboard/dashboard.py:462]: REMOVE!!!!!!!

File: dashboard/dashboard.py [hunk 4]
```diff
@@ -13,9 +13,8 @@
 from email_validator import validate_email as ve, EmailNotValidError
 import phonenumbers
 from phonenumbers import NumberParseException
-import numpy as np
```
Comment (olliejth) [inline dashboard/dashboard.py:16]: Nice naming.

File: dashboard/dashboard.py [hunk 5]
```diff
@@ -125,6 +124,8 @@ def validate_phone_number(phone_number:str, region='GB') -> bool:
             filtered_df['cloud_coverage_rolling'] = filtered_df['cloud_coverage_percent'].rolling(window=24).mean()
             filtered_df['visibility_rolling'] = filtered_df['visibility_m'].rolling(window=24).mean()
             filtered_df['temperature_rolling'] = filtered_df['temperature_c'].rolling(window=24).mean()
+            filtered_df['precipitation_prob_rolling'] = filtered_df['precipitation_probability_percent'].rolling(window=24).mean()
+            filtered_df['precipitation_mm_rolling'] = filtered_df['precipitation_mm'].rolling(window=24).mean()
```
Comment (olliejth) [inline dashboard/dashboard.py:128]: Nice use of rolling.

File: dashboard/dashboard.py [hunk 6]
```diff
@@ -212,12 +213,47 @@ def validate_phone_number(phone_number:str, region='GB') -> bool:
...
+        )
+
+        # Precipitation amount (mm) on the right y-axis.
+        precipitation_mm_chart = alt.Chart(filtered_df).mark_line(color='green').encode(
+            x=alt.X('at:T', title='Time'),
+            y=alt.Y('precipitation_mm:Q', title='Precipitation (mm)'),
+            tooltip=['at:T', 'precipitation_mm']
+        )
+
+        precipitation_chart = alt.layer(
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:237]: Love layering

**Review by AlexOBarnes — APPROVED** [review]
thanks for making those changes

File: dashboard/hold.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        x=alt.X('x:O', title='Date', axis=alt.Axis(labels=False, ticks=False)),
+        y=alt.Y('y:O', title='Celestial Body'),
+        color=alt.Color('z:Q',
+                        scale=alt.Scale(domain=[0, 30000],
+                                        range=['#fff5eb', '#ff0000', '#800080']),
+                        title='Visibility (m)'),
+        tooltip=['y', 'x:T', 'z']
+    ).properties(
+        width=2000,
+        height=500,
```
Comment (olliejth) [inline dashboard/hold.py:47]: Good idea to specify this 🥇

File: dashboard/load_dashboard_data.py (hunk 2 above)
Comment (olliejth) [reply dashboard/load_dashboard_data.py:98]: Aaaaarrgghhh 😨

File: dashboard/load_dashboard_data.py [hunk 8]
```diff
@@ -69,13 +71,95 @@ def load_forecasts_by_county_name(conn:Psycopg2Connection) -> pd.DataFrame:
...
+        tqdm.pandas()  
+        bodies_df['visibility_coefficient'] = bodies_df.progress_apply(
+            lambda row: starwatch_coefficient(row['visibility_m'], row['distance_km'], row['cloud_coverage_percent'], k=0.01),
+            axis=1
+        )
+
+        print(bodies_df.tail(10))
+        bodies_df.to_csv('test.csv')
+    else:
+        print('No data fetched. Exiting.')
```
Comment (olliejth) [inline dashboard/load_dashboard_data.py:165]: Exciting

Comment (olliejth) [reply dashboard/load_dashboard_data.py:165]: This should be done using logging. 🌴

File: dashboard/requirements.txt [hunk 9]
```diff
@@ -10,4 +10,5 @@ requests
 python-dotenv
 matplotlib
 numpy
-requests
\ No newline at end of file
+requests
```
Comment (olliejth) [inline dashboard/requirements.txt:13]: Nice replacement.

**Review by olliejth — APPROVED** [review]
Large sections of commented out code. Logging should be implemented instead of print statements.

**Pull request merged**